### DISCUSSION TITLE
Wildcard hostname with Synology DDNS

### DISCUSSION THREAD

**Discussion by leesherwood** [willswire/unifi-ddns#93] [Ideas]
_Due to ISP restrictions, I have to use a double NAT setup whereby my ISP router has as much disabled as possible (firewall, wifi, etc) and my UDM is placed in it's DMZ so all incoming requests get forwarded on to it, where the UDM's firewall & routing can deal with what to do with it. This means my UDM doesn't see the true WAN IP, it see's it's IP assigned from the ISP router. As there's no option to have the UDM discover the WAN IP from an external source, I can't use it._

_I have a Synology NAS with DDNS support, which does discover the true WAN IP, but it doesn't support multiple hostnames or wildcards. So I modified the script slightly to support a `wildcard` param to get around that. I figured I'd share it here in case anyone else find's it useful._

-------

1. Modify the script to look for a `wildcard` parameter, when detected it will append a `*.hostname` entry to the list of hostnames to be updated. 

```diff
diff --git a/src/index.js b/src/index.js
index ece4843..bd82ff4 100644
--- a/src/index.js
+++ b/src/index.js
@@ -122,6 +122,12 @@ async function handleRequest(request) {
        const hostnameParam = params?.get("hostname") || params?.get("host") || params?.get("domains");
        const hostnames = hostnameParam?.split(",");
 
+       // Synology doesn't allow using the same DDNS provider twice, and doesn't accept a CSV hostname
+       // So I added a ?wildcard=1 parameter
+       if (params?.get('wildcard') && parseInt(params.get('wildcard'), 10) === 1) {
+               hostnames.forEach( h => hostnames.push('*.'+h));
+       }
+
        // fallback to connecting IP address
        const ipsParam = params.get("ips") || params.get("ip") || params.get("myip") || request.headers.get("Cf-Connecting-Ip");
        const ips = ipsParam?.split(",");
```

2. Deploy this to CloudFlare via `npx wrangler deploy` just like the original. Alternatively, you can edit the code directly in Cloudflare if you originally used the deploy button.

3. Create a new custom provider in Synology DDNS

Service Provider: `CloudflareWorker` (use whatever you like, it's just a reference name)
Query URL: `https://yourWorkerUrl.workers.dev/update?ip=__MYIP__&hostname=__HOSTNAME__&wildcard=1`

4. Add the DDNS provider

Service Provider: `USER_CloudflareWorker`
Hostname: `mydomain.com`
Username: `mydomain.com`
Password: `yourApiToken`

**Comment by willswire** [comment]
Thanks for sharing this with the community!
